question-servers/freeform.js

```diff
@@ -149,6 +151,101 @@ module.exports = {
+                async.map(elementFolders, (element) => {
+                    fs.readdir(path.join(sourceDir, element), (err, extensions) => {
+                        if (ERR(err, callback)) return;
+                        return callback(null, extensions.map(ext => [element, ext]));
+                    });
+                }, callback);
+            },
+            (extensions, callback) => {
+                /* Filter out non-directories */
+                async.filter(extensions, (extension, callback) => {
+                    const [element, extensionDir] = extension;
+                    fs.lstat(path.join(sourceDir, element, extensionDir), (err, stats) => {
+                        if (ERR(err, callback)) return;
+                        callback(null, stats.isDirectory());
+                    });
+                }, (err, results) => {
+                    if (ERR(err, callback)) return;
+                    return callback(null, results);
+                });
+            },
+            (extensions, callback) => {
+                const elements = {};
+
+                /* Populate element map */
+                extensions.forEach((extension) => {
+                    if (!(extension[0] in elements)) {
+                        elements[extension[0]] = {};
+                    }
+                });
+
+                /* Load extensions */
+                async.each(extensions, (extension, callback) => {
+                    const [element, extensionDir] = extension;
+                    const infoPath = path.join(sourceDir, element, extensionDir, 'info.json');
+                    fs.readJson(infoPath, (err, info) => {
+                        if (err && err.code === 'ENOENT') {
+                            /* Not an extension directory, skip it. */
+                            logger.verbose(`${infoPath} not found, skipping...`);
+                            return callback(null);
+                        }
+                        if (ERR(err, callback)) return;
+
+                        jsonLoader.validateJSON(info, schemas.infoElementExtension, (err) => {
+                            if (ERR(err, callback)) return;
+                            info.name = extensionDir;
+                            info.directory = path.join(sourceDir, element, extensionDir);
+                            elements[element][extensionDir] = info;
+                            callback(null);
+                        });
+                    });
+                }, (err) => {
+                    if (ERR(err, callback)) return;
+                    callback(null, elements);
+                });
+            },
+        ], (err, extensions) => {
+            if (ERR(err, callback)) return;
+            return callback(null, extensions);
+        });
+    },
+
+    loadExtensionsForCourse(course, callback) {
+        if (courseExtensionsCache[course.id] !== undefined) {
+            return callback(null, courseExtensionsCache[course.id]);
+        }
```

Comment:
We should key this caching by `course.commit_hash`, so that we automatically reload whenever the `commit_hash` changes. At the moment we need to explicitly called `reloadExtensionsForCourse` in `syncFromDisk.js`, which is bad because it only happens on the server instance that actually did the sync. Alternatively, we could send a message via redis to tell the other servers to reload, but it seems simpler and more robust to just key on `commit_hash`.

Just to be clear, this is not a new problem with element extensions. The existing element cache also has this problem, as far as I'm aware.

Reply:
In that case, would we have to worry about periodically emptying out or pruning the caches?  Maybe this isn't that big of a deal(?) but I could see all the old commit hash cache entries wasting memory here.

Reply:
I guess one simple strategy would be to set `courseExtensionsCache[course.id] = {commit_hash: course.commit_hash, data: actualCachedData}`. Then rather than simply checking whether `courseExtensionsCache[course.id] !== undefined`, we would also check whether the `commit_hash` matched to figure out whether we need to use the cache or update it. When we do update it, we'd overwrite the old one so we don't get growth.